File: problems/0268_missing-number/solution_xor.py
from typing import List, Optional


class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        # Seed with n — the one index the loop below never visits — then
        # fold every index 0..n-1 and every element into one accumulator.
        result = len(nums)
        for i, value in enumerate(nums):
            # Each present value matches an index and cancels it; the absent
            # value pairs with nothing and survives the fold.
            result ^= i ^ value
        return result

File: problems/0200-0300/0268_missing-number/solution_xor.cpp
class Solution {
  public:
    int missingNumber(vector<int> &nums) {
        // Seed with n — the one index the loop below never visits — then
        // fold every index 0..n-1 and every element into one accumulator.
        int result = nums.size();
        for (int i = 0; i < (int)nums.size(); ++i) {
            // Each present value matches an index and cancels it; the absent
            // value pairs with nothing and survives the fold.
            result ^= i ^ nums[i];
        }
        return result;
    }
};
